src/mongoose_batch_worker.erl

```diff
@@ -0,0 +1,136 @@
+-spec init({mongooseim:host_type(),
+            pos_integer(),
+            pos_integer(),
+            fun((list(), map()) -> ok | {error, term()}),
+            map()}) -> {ok, state()}.
+init({HostType, Interval, MaxSize, FlushCallback, FlushExtra}) ->
+    ?LOG_INFO(#{what => starting_worker, host_type => HostType, pool_id => FlushCallback}),
+    {ok, #state{host_type = HostType,
+                batch_size = MaxSize,
+                flush_interval = Interval,
+                flush_callback = FlushCallback,
+                flush_extra = FlushExtra}}.
+
+-spec handle_call(term(), {pid(), term()}, state()) -> {reply, term(), state()}.
+handle_call(Msg, From, State) ->
+    ?UNEXPECTED_CALL(Msg, From),
+    {reply, unexpected_call, State}.
+
+-spec handle_cast(task | term(), state()) -> {noreply, state()}.
+handle_cast({task, Task}, State) ->
+    {noreply, handle_task(Task, State)};
+handle_cast(Msg, State) ->
+    ?UNEXPECTED_CAST(Msg),
+    {noreply, State}.
+
+-spec handle_info(flush | term(), state()) -> {noreply, state()}.
+handle_info(flush, State) ->
+    {noreply, run_flush(State)};
+handle_info({garbage_collect, asynchronous_gc_triggered, true}, State) ->
+    {noreply, State};
+handle_info(Msg, State) ->
+    ?UNEXPECTED_INFO(Msg),
+    {noreply, State}.
+
+-spec terminate(term(), state()) -> term().
+terminate(_Reason, State) ->
+    ?LOG_INFO(log_fields(State, #{what => stopping_worker})),
+    case [] =:= State#state.flush_queue of
+        true -> ok;
+        false ->
+            do_run_flush(State),
+            ?LOG_WARNING(log_fields(State, #{what => worker_terminate_requires_flush}))
+    end.
+
+-spec code_change(term(), state(), term()) -> {ok, state()}.
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+% Don't leak the tasks to logs, can contain private information
+format_status(_Opt, [_PDict, State | _]) ->
+    [{data, [{"State", State#state{flush_queue = []}},
+             {"Task Queue Length", length(State#state.flush_queue)}]}].
+
+%% Batched tasks callbacks
+handle_task(Task, State0 = #state{batch_size = MaxSize,
+                                  flush_interval = Interval,
+                                  flush_interval_tref = TRef0,
+                                  flush_queue = Acc0}) ->
+    Length = length(Acc0),
+    TRef1 = maybe_schedule_flush(TRef0, Length, Interval),
+    State1 = State0#state{flush_interval_tref = TRef1,
+                          flush_queue = [Task | Acc0]},
+    case Length + 1 >= MaxSize of
+        false -> State1;
+        true -> run_flush(State1)
+    end.
+
+maybe_schedule_flush(undefined, 0, Interval) ->
+    erlang:send_after(Interval, self(), flush);
+maybe_schedule_flush(TRef, _, _) ->
+    TRef.
+
+run_flush(State = #state{flush_queue = Queue, flush_interval_tref = TRef}) ->
+    cancel_and_flush_timer(TRef),
+    ?LOG_DEBUG(log_fields(State, #{what => worker_flush})),
+    NewState = do_run_flush(State#state{flush_queue = Queue,
```

Comment:
`flush_queue = Queue,` does nothing here

Reply:
Ah, probably some remnant of a time when it actually did something xD